[documentation] Update skaffold & docker-compose local install (#110)
SubstraFoundation/substra: docs/local_install_skaffold.md

Once installed, launch Docker and open its "preferences" panel. In the Kubernetes tab, check the `Enable Kubernetes` checkbox.

Kubernetes will take a while to launch the first time, but once it is done, you can move on to configuring.

If you don't have `homebrew` installed, please install it first: <https://brew.sh/>

##### Ubuntu

Please use **[Kubernetes v1.15](https://kubernetes.io/docs/tasks/tools/install-kubectl/)** (client & server):

```sh
curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.15.0/bin/linux/amd64/kubectl
chmod +x ./kubectl
sudo mv ./kubectl /usr/local/bin/kubectl
kubectl version --client
```

#### [2. Minikube](https://minikube.sigs.k8s.io/docs/start/)

```sh
# Ubuntu/Debian only
# Get the executable & install it (Please use the up-to-date version)
curl -LO https://storage.googleapis.com/minikube/releases/latest/minikube_1.6.2.deb && sudo dpkg -i minikube_1.6.2.deb
```

#### [3. Helm](https://helm.sh/)

V3 is not supported yet, please use [Helm v2.16.1](https://github.com/helm/helm/releases/tag/v2.16.1)

```sh
# Mac
# If not already installed with Docker Desktop:

> brew install kubernetes-helm

Author: Thank you!

    - [Start Substra](#start-substra)
    - [Stop Substra](#stop-substra)
    - [Reset Substra](#reset-substra)
  - [4. Login, password and urls](#4-login-password-and-urls)
    - [Credentials and urls](#credentials-and-urls)
    - [Browser extension](#browser-extension)
    - [Substra CLI config & login](#substra-cli-config--login)
  - [5. Troubleshooting](#5-troubleshooting)
    - [Virtualization issues](#virtualization-issues)
    - [Kubectl useful commands](#kubectl-useful-commands)
    - [Minikube useful commands](#minikube-useful-commands)
    - [Tiller](#tiller)
    - [[WIP] Ongoing issues](#wip-ongoing-issues)
  - [6. Need help?](#6-need-help)
  - [7. Further resources](#7-further-resources)
  - [8. Acknowledgements](#8-acknowledgements)

___

When everything is ready, you will be able to start Substra with:

```sh
# Ubuntu only
minikube start --cpus 6 --memory 8192 --disk-size 50g --kubernetes-version='v1.15.4'

# In 3 different terminal windows, in this order:
# In the repository hlf-k8s
skaffold dev

# In the repository susbtra-backend
skaffold dev

# In the repository substra-frontend
skaffold dev
```

## 1. Substra Setup

### General knowledge

> I order to install Substra, it is *recommended* to be comfortable with your package manager and to have some basic knowledge of Unix (Mac or Ubuntu/Debian) administration and network.

Reviewer: I would add Kubernetes basic knowledge as well?

Author: You're right, thank you!

### Hardware requirements

> Please be cautious with the parameters passed to Kubernetes as it might try to use all the allocated resources without regards for your system!

If you wish to comfortably run Substra, it is advised to have:

- **50 GB of free space** as several images need to be pulled.
- **8 GB of RAM** for Kubernetes.

### Software requirements

> Note: Please **always** refer to the package provider website before installing any software!
> Note: This setup has been tested on Ubuntu 19.10, 18.04.3 & Linux Mint 19 Cinnamon (4.15.0-45-generic Kernel)

> Substra deployment is orchestrated by `Kubernetes` and `Minikube` is a great tool for your local Kubernetes deployments.

Reviewer: What about splitting the sentence in 2? The first part is generic and the second part targets more Linux users. We could also recommend `Docker for Mac with Kubernetes` for mac users?

Author: Agreed, but Mac users can use minikube too! ;)

`Helm` is a Kubernetes package manager that helps install and manage Kubernetes applications. `Tiller` runs inside the Kubernetes Cluster and manages releases of your charts. Helm has two parts : helm (client) & tiller (server). Charts are Helm packages that contains a package description (`Chart.yaml`) and one or more templates which contain Kubernetes manifest files. Charts can be stored locally or on a distant repository.

`Skaffold` is a program working on top of the Kubernetes configuration that will operate the deployment for you. It relies on the `skaffold.yaml` files that you will find at the root of each repositories.

> #### [1. Kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)

Reviewer: The title `Kubectl` seems incorrect with the content (at least for mac users).
 
What about refactoring the first 2 sections into:

1. Running Kubernetes locally
- Ubuntu: minikube
- Mac: Docker for mac

2. Installing Kubectl
- Ubuntu:  `apt-get`
- Mac: `brew install kubectl`

```sh
helm init
# or
helm init --upgrade
# you might need to use
helm init --service-account tiller --upgrade
```

#### Network

- Ubuntu

```sh
# Append your Kubernetes cluster ip to your system hosts
echo "$(minikube ip) substra-backend.node-1.com substra-frontend.node-1.com substra-backend.node-2.com substra-frontend.node-2.com" | sudo tee -a /etc/hosts

# Inside a VM, you will need to use
echo "$(sudo minikube ip) substra-backend.node-1.com substra-frontend.node-1.com substra-backend.node-2.com substra-frontend.node-2.com" | sudo tee -a /etc/hosts
```

Example:

```sh
192.168.39.32 substra-backend.node-1.com substra-frontend.node-1.com substra-backend.node-2.com substra-frontend.node-2.com
```

If you want to customize your configuration,
you can assign the ingress loadbalancer ip to the corresponding url, for example:

```sh
10.111.206.49 substra-backend.node-1.com substra-frontend.node-1.com
10.109.77.31 substra-backend.node-2.com substra-frontend.node-2.com
```

If you want to expose another service, you can use something like:

```sh
kubectl port-forward -n ${NAMESPACE} service/${SERVICE_NAME} ${SERVICE_PORT}:${PORT_WANTED}
```

- Mac

> Once done, we can configure your machine so that the urls `substra-backend.node-1.com`, `substra-backend.node-2.com`, `substra-frontend.node-1.com` and `substra-frontend.node-2.com` point to your local instance of the platform.

Reviewer: It seems to be valid for both Ubuntu and Mac. Maybe it could be moved at the top of Network part?

Author: Bien vu, merci @samlesu

Author: Sorry for the french moment, thanks for your input @samlesu

curl -LO https://storage.googleapis.com/minikube/releases/latest/minikube_1.6.2.deb && sudo dpkg -i minikube_1.6.2.deb
```

#### [3. Helm](https://helm.sh/)

V3 is not supported yet, please use [Helm v2.16.1](https://github.com/helm/helm/releases/tag/v2.16.1) to get Helm and Tiller.

```sh
# Mac
# If not already installed with Docker Desktop:
brew install kubernetes-helm

# Linux (amd64)
# Get the executable
curl -LO https://get.helm.sh/helm-v2.16.1-linux-amd64.tar.gz
# Extract the downloaded archive
tar xzvf helm-v2.16.1-linux-amd64.tar.gz
cd linux-amd64/
# Move the executables to your local bin
sudo mv tiller helm /usr/local/bin/
```

#### [4. Skaffold](https://skaffold.dev/)

```sh
# Mac
brew install skaffold

# Linux
# Get the executable
curl -Lo skaffold https://storage.googleapis.com/skaffold/releases/latest/skaffold-linux-amd64
# Make it executable on your machine
chmod +x skaffold
# Move it to your local bin
sudo mv skaffold /usr/local/bin
```

### Virtualisation

- If you are using [VirtualBox](https://www.virtualbox.org/wiki/Downloads) with Ubuntu, you will have to:
  - install `socat` with the command `sudo apt-get install socat`
  - launch all commands with `sudo`

>   - pass the parameter `--vm-driver=none` when starting Minikube (`minikube start...`) 

Author: Ok, good point, thank you! I was thinking about starting Kubernetes inside a VM

# Local install of Substra using Kubernetes and Skaffold

> This is an ongoing document, please feel free to reach us or to raise any issue.

> This guide will help you to run the Substra platform on your machine in development mode, with a two organisations setup.

Reviewer 2: typo: organizations

Also, we don't use the word organization anywhere else, we tend to use "node".

Author: Noted!

  - [2. Setup](#2-setup)
    - [1. Get the source code](#1-get-the-source-code)
    - [2. Configuration](#2-configuration)
      - [Minikube](#minikube)
      - [Network](#network)
        - [Ubuntu](#ubuntu-1)
        - [Mac](#mac-1)
  - [3. Running the platform](#3-running-the-platform)
    - [Start Substra](#start-substra)
    - [Stop Substra](#stop-substra)
  - [4. Login, password and urls](#4-login-password-and-urls)
    - [Credentials and urls](#credentials-and-urls)
    - [Browser extension](#browser-extension)
    - [Substra CLI config & login](#substra-cli-config--login)
  - [5. Titanic example](#5-titanic-example)
  - [6. Troubleshooting](#6-troubleshooting)
    - [Virtualisation](#virtualisation-1)
    - [Kubectl useful commands](#kubectl-useful-commands)
    - [Minikube useful commands](#minikube-useful-commands)
    - [Tiller](#tiller)
    - [Ongoing issues](#ongoing-issues)
  - [7. Further resources](#7-further-resources)
  - [8. Need help](#8-need-help)
  - [9. Acknowledgements](#9-acknowledgements)
___

When everything is ready, you will be able to start Substra with:
```sh
minikube start --cpus 6 --memory 8192 --disk-size 50g --kubernetes-version='v1.15.4'

> skaffold dev # in hlf-k8s, susbtra-backend, substra-frontend repositories

Reviewer 2: You could make this even more explicit by telling users to run `skaffold dev` in all 3 directories in independent shells otherwise they may wait for the first one to finish to launch the next one.

Author: You're right!

```

## 1. Requirements

### General knowledge

> I order to install Substra, it is *recommended* to be confortable with your package manager and to have some basic knowledge of Linux administration and network.

Reviewer 2: typo: comfortable

### Hardware requirements

> Please be cautious with the parameters passed to Kubernetes as it might try to use all the allocated resources without regards for your system!

If you wish to comfortably run Substra, it is advised to have:

- At least 30 Go of free space but **50** would really be better!

> - **8 Go of RAM** for Kubernetes/Minikube! Be warned, 4 Go won't work. 

Reviewer 2: typo: GB

`Skaffold` is a program working on top of the Kubernetes configuration that will operate the deployment for you. It relies on the `skaffold.yaml` files that you will find at the root of each repositories.

#### [1. Kubernetes](https://kubernetes.io/docs/tasks/tools/install-kubectl/)

##### Mac

First of all, download the `Docker desktop` installer from https://www.docker.com/products/docker-desktop. You'll have to create an account there to do so. Then run it to install Docker on your machine.

Once installed, launch Docker and open its "preferences" panel. In the Kubernetes tab, check the `Enable Kubernetes` checkbox.

Kubernetes will take a while to launch the first time, but once it is done, you can move on to configuring.

If you don't have `homebrew` installed, please install it first with the following command:

```sh
/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"

> ```

Reviewer 2: Instead of duplicating the install instructions from homebrew's website, why not send the reader there ?

Author: Sure!

##### Ubuntu

Please use **[Kubernetes v1.15](https://kubernetes.io/docs/tasks/tools/install-kubectl/)** (client & server):

```sh
curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.15.0/bin/linux/amd64/kubectl
chmod +x ./kubectl
sudo mv ./kubectl /usr/local/bin/kubectl
kubectl version --client
```

#### [2. Minikube](https://minikube.sigs.k8s.io/docs/start/)

```sh
# Mac

> brew install minikube

Reviewer 2: Are you sure we want to direct mac users to minikube ? The bundled k8s is much easier to use.

# Ubuntu
# Get the executable & install it (Please use the up-to-date version)
curl -LO https://storage.googleapis.com/minikube/releases/latest/minikube_1.6.2.deb && sudo dpkg -i minikube_1.6.2.deb
```

#### [3. Helm](https://helm.sh/)

V3 is not supported yet, please use [Helm v2.16.1](https://github.com/helm/helm/releases/tag/v2.16.1)

```sh
# Mac
# Get the executable
curl -LO https://get.helm.sh/helm-v2.16.1-darwin-amd64.tar.gz
# Extract the downloaded archive
tar xzvf helm-v2.16.1-darwin-amd64.tar.gz
cd darwin-amd64/
# Move the executables to your local bin

> sudo mv tiller helm /usr/local/bin/

Reviewer 2: I think you can use a single brew command for this.

# Linux (amd64)
# Get the executable
curl -LO https://get.helm.sh/helm-v2.16.1-linux-amd64.tar.gz
# Extract the downloaded archive
tar xzvf helm-v2.16.1-linux-amd64.tar.gz
cd linux-amd64/
# Move the executables to your local bin
sudo mv tiller helm /usr/local/bin/
```

#### [4. Skaffold](https://skaffold.dev/)

```sh
# Mac
brew install skaffold
# Or get the executable
curl -Lo skaffold https://storage.googleapis.com/skaffold/releases/latest/skaffold-darwin-amd64
# Make it executable on your machine
chmod +x skaffold
# Move it to your local bin

> sudo mv skaffold /usr/local/bin

Reviewer 2: Since we tell the user to have homebrew installed, we could just use the brew command and not specify an alternate one.

  - install `socat` with the command `sudo apt-get install socat`
  - launch all commands with `sudo`
  - pass the parameter `--vm-driver=none` when starting Minikube (`minikube start...`) 
- Ubuntu:
  - Validate your host virtualisation with the command `virt-host-validate`: <https://linux.die.net/man/1/virt-host-validate>
  - [KVM (Kernel Virtual Machine) installation](https://help.ubuntu.com/community/KVM/Installation#Installation)
  - Required packages: [Ubuntu help](https://help.ubuntu.com/community/KVM/Installation#Install_Necessary_Packages)
  - If you need more information about [libvirt & qemu](https://libvirt.org/drvqemu.html)

## 2. Setup

### 1. Get the source code

You will find the main Substra repository [here](https://github.com/SubstraFoundation/substra), but in order to run the Substra framework, you will need to clone 3 repositories: [hlf-k8s](https://github.com/SubstraFoundation/hlf-k8s) (Hyperledger Fabric), [susbtra-backend](https://github.com/SubstraFoundation/substra-backend) and [substra-frontend](https://github.com/SubstraFoundation/substra-frontend).

> The `hlf-k8s` repository is in charge of the initialisation of the [Hyperledger Fabric](https://www.hyperledger.org/projects/fabric) network. By default, it will create an `orderer` and two orgs (`org-1` & `org-2`).

Reviewer 2: typo: initialization

RepoToClone=(
# https
https://github.com/SubstraFoundation/substra.git
https://github.com/SubstraFoundation/hlf-k8s.git
https://github.com/SubstraFoundation/substra-frontend.git
https://github.com/SubstraFoundation/substra-backend.git

# ssh
# git@github.com:SubstraFoundation/substra.git
# git@github.com:SubstraFoundation/hlf-k8s.git
# git@github.com:SubstraFoundation/substra-backend.git
# git@github.com:SubstraFoundation/substra-frontend.git
)

for repo in ${RepoToClone[@]}
do
    echo "Cloning" $repo
    git clone $repo
done
```

TODO: add script?

> Note: if you do not have git on your machine, you can also download and unzip the code using these links:
> - [hlf-k8s](https://github.com/SubstraFoundation/hlf-k8s/archive/master.zip)
> - [substra-backend](https://github.com/SubstraFoundation/substra-backend/archive/master.zip)
> - [substra-frontend](https://github.com/SubstraFoundation/substra-frontend/archive/master.zip)

### 2. Configuration

#### Minikube

Please enable the ingress minikube module: `minikube addons enables ingress`. You might need to edit `skaffold.yaml` files and set `nginx-ingress.enabled` to `false`.

You can now start Minikube with:

```sh
cd hlf-k8s

> # Confortable setup

Reviewer 2: typo: comfortable

minikube start --cpus 6 --memory 8192 --disk-size 50g --kubernetes-version='v1.15.4'
# Frugal setup
minikube start --cpus 4 --memory 8192 --disk-size 30g --kubernetes-version='v1.15.4'
# VM setup (Please note that inside the VM, you will to execute all commands with sudo)
sudo minikube start --vm-driver=none --kubernetes-version='v1.15.4'

# first time only
helm init
# or you might need to 
helm init --service-account tiller --upgrade
```

#### Network

##### Ubuntu

```sh
# Append your Kubernetes cluster ip to your system hosts
echo "$(minikube ip) substra-backend.node-1.com substra-frontend.node-1.com substra-backend.node-2.com substra-frontend.node-2.com" >> /etc/hosts

# Inside a VM you will need to use
echo "$(sudo minikube ip) substra-backend.node-1.com substra-frontend.node-1.com substra-backend.node-2.com substra-frontend.node-2.com" | sudo tee -a /etc/hosts
```

Example:
```sh
192.168.39.32 substra-backend.node-1.com substra-frontend.node-1.com substra-backend.node-2.com substra-frontend.node-2.com
```

> If you want to customise your configuration,

Reviewer 2: typo: customize

# Local installation of Substra using Kubernetes and Skaffold

> This is an ongoing document, please feel free to reach us or to raise any issue.

This guide will help you to run the Substra platform on your machine in development mode, with a two nodes setup.

- [Local installation of Substra using Kubernetes and Skaffold](#local-installation-of-substra-using-kubernetes-and-skaffold)
  - [1. Requirements](#1-requirements)
    - [General knowledge](#general-knowledge)
    - [Hardware requirements](#hardware-requirements)
    - [Software requirements](#software-requirements)
      - [1. Kubernetes](#1-kubernetes)
        - [Mac](#mac)
        - [Ubuntu](#ubuntu)
      - [2. Minikube](#2-minikube)
      - [3. Helm](#3-helm)
      - [4. Skaffold](#4-skaffold)
    - [Virtualization](#virtualization)
  - [2. Setup](#2-setup)
    - [1. Get the source code](#1-get-the-source-code)

>     - [2. Configuration](#2-configuration)

Reviewer 3: Wouldn't be better to bundle this part with the sections above ?

If you wish to comfortably run Substra, it is advised to have:

- **50 GB of free space** as several images need to be pulled.
- **8 GB of RAM** for Kubernetes.

### Software requirements

> Note: Please **always** refer to the package provider website before installing any software!
> Note: This setup has been tested on Ubuntu 19.10, 18.04.3 & Linux Mint 19 Cinnamon (4.15.0-45-generic Kernel)

TODO: ADD PYTHON 3 minimum version
TODO: ADD python version to venv

Substra deployment is orchestrated by `Kubernetes` and `Minikube` is a great tool for your local Kubernetes deployments.

`Helm` is a Kubernetes package manager that helps install and manage Kubernetes applications. `Tiller` runs inside the Kubernetes Cluster and manages releases of your charts. Helm has two parts : helm (client) & tiller (server). Charts are Helm packages that contains a package description (`Chart.yaml`) and one or more templates which contain Kubernetes manifest files. Charts can be stored locally or on a distant repository.

The general file structure is presented on Helm's [website](https://v2.helm.sh/docs/developing_charts/#the-chart-file-structure):

> ```yaml

Reviewer 3: I would not put this here, I would rather let people go on the doc of helm

#### [1. Kubernetes](https://kubernetes.io/docs/tasks/tools/install-kubectl/)

##### Mac

First of all, download the `Docker desktop` installer from <https://www.docker.com/products/docker-desktop>. You'll have to create an account there to do so. Then run it to install Docker on your machine.

Once installed, launch Docker and open its "preferences" panel. In the Kubernetes tab, check the `Enable Kubernetes` checkbox.

Kubernetes will take a while to launch the first time, but once it is done, you can move on to configuring.

If you don't have `homebrew` installed, please install it first: <https://brew.sh/>

##### Ubuntu

Please use **[Kubernetes v1.15](https://kubernetes.io/docs/tasks/tools/install-kubectl/)** (client & server):

```sh
curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.15.0/bin/linux/amd64/kubectl
chmod +x ./kubectl
sudo mv ./kubectl /usr/local/bin/kubectl
kubectl version --client
```

#### [2. Minikube](https://minikube.sigs.k8s.io/docs/start/)

> 

Reviewer 3: Maybe add something explaining that on mac, docker-desktop can be used ?

sudo minikube start --vm-driver=none --kubernetes-version='v1.15.4'
```

#### Helm init

The first time you install Substra, you will need to use:

```sh
helm init
# or
helm init --upgrade
# you might need to use
helm init --service-account tiller --upgrade
```

#### Network

- Ubuntu

```sh
# Append your Kubernetes cluster ip to your system hosts
echo "$(minikube ip) substra-backend.node-1.com substra-frontend.node-1.com substra-backend.node-2.com substra-frontend.node-2.com" | sudo tee -a /etc/hosts

# Inside a VM, you will need to use
echo "$(sudo minikube ip) substra-backend.node-1.com substra-frontend.node-1.com substra-backend.node-2.com substra-frontend.node-2.com" | sudo tee -a /etc/hosts
```

Example:

```sh
192.168.39.32 substra-backend.node-1.com substra-frontend.node-1.com substra-backend.node-2.com substra-frontend.node-2.com
```

If you want to customize your configuration,
you can assign the ingress loadbalancer ip to the corresponding url, for example:

```sh
10.111.206.49 substra-backend.node-1.com substra-frontend.node-1.com
10.109.77.31 substra-backend.node-2.com substra-frontend.node-2.com
```

> If you to expose another service, you can use something like:

Reviewer 4: I think you missed a word here 🙂

    - [Stop Substra](#stop-substra)
    - [Reset Substra](#reset-substra)
  - [Login, password and urls](#login-password-and-urls)
    - [Credentials and urls](#credentials-and-urls)
    - [Browser extension](#browser-extension)
    - [Substra CLI config & login (Mac & Ubuntu)](#substra-cli-config--login-mac--ubuntu)
  - [Troubleshooting](#troubleshooting)
    - [Virtualization issues](#virtualization-issues)
    - [Kubectl useful commands](#kubectl-useful-commands)
    - [Minikube useful commands](#minikube-useful-commands)
    - [Tiller](#tiller)
    - [[WIP] Ongoing issues](#wip-ongoing-issues)
  - [Need help?](#need-help)
  - [Further resources](#further-resources)
  - [Acknowledgements](#acknowledgements)

___

When everything is ready, you will be able to start Substra with:

```sh
# If you use Minikube
minikube start --cpus 6 --memory 8192 --disk-size 50g --kubernetes-version='v1.15.4'

# In 3 different terminal windows, in this order:
# In the repository hlf-k8s
skaffold dev

# In the repository susbtra-backend
skaffold dev

# In the repository substra-frontend
skaffold dev
```

## Substra Setup

### General knowledge

> In order to install Substra, it is *recommended* to be comfortable with your package manager and to have some basic knowledge of Unix (Mac or Ubuntu/Debian) administration and network. It might also be useful to have a good command of Docker containers and Kubernetes orchestration.

Reviewer: ```suggestion
In order to install Substra, it is *recommended* to be comfortable with your package manager and to have some basic knowledge of Unix (Mac or Ubuntu/Debian) administration and network. It might also be useful to have a good knowledge of Docker containers and Kubernetes orchestration.
```

echo "192.168.65.2 substra-backend.node-1.com substra-frontend.node-1.com substra-backend.node-2.com substra-frontend.node-2.com" | sudo tee -a /etc/hosts
```

## 3. Running the platform

### Start Substra

- For Ubuntu, once Minikube is running and Tiller initialized, please go the `hlf-k8s`, `substra-backend` and `substra-frontend` repositories and run `skaffold dev` to start Substra with Skaffold.

> - For Mac, directly head to the tree repositories and start Substra with Skaffold.

Reviewer: I think we could avoid the distinction between Ubuntu and Mac here: both setups require a Kubernetes cluster up and running (started via Minikube or Docker for mac) and Tiller initialized.

Author: Done, thank you!

echo "192.168.65.2 substra-backend.node-1.com substra-frontend.node-1.com substra-backend.node-2.com substra-frontend.node-2.com" | sudo tee -a /etc/hosts
```

## 3. Running the platform

### Start Substra

> Once Minikube is running and Tiller initialised, please go the `hlf-k8s`, `substra-backend` and `substra-frontend` repositories and run `skaffold dev` to start Substra with Skaffold:

Reviewer 2: typo: initialized

# or
[network-org-2-peer-1-hlf-k8s-channel-join-0-kljgq fabric-tools] 2020-02-10 10:18:02.350 UTC [channelCmd] executeJoin -> INFO 002 Successfully submitted proposal to join channel
```

![status:200 payload:"OK"](/substra/docs/img/start_hlf-k8s.png "status:200 payload:'OK'")

#### 2. substra-backend repository

In the `substra-backend` repository, please run the command `skaffold dev`. The platform will be ready once the terminal displays:

```sh
[backend-org-2-substra-backend-server-74bb8486fb-nkq6m substra-backend] INFO - 2020-02-10 10:24:42,514 - django.server - "GET /liveness HTTP/1.1" 200 2
# or
[backend-org-1-substra-backend-server-77cf8cb9fd-cwgs6 substra-backend] INFO - 2020-02-10 10:24:51,393 - django.server - "GET /readiness HTTP/1.1" 200 2
```

![django.server readiness](/substra/docs/img/start_backend.png "django.server readiness")

#### 3. substra-frontend repository

In the `susbtra-frontend` repository, please run the command `skaffold dev`. The platform will be ready once the terminal displays:

```sh
[frontend-org-2-substra-frontend-787554fc4b-pmh2g substra-frontend] CACHING:  /login
```

![CACHING Login](/substra/docs/img/start_frontend.png "CACHING Login")

> Alternatively, instead of using `skaffold`, you might want to start the `substra-frontend` with [Yarn](https://yarnpkg.com/getting-started/install):

Reviewer: I think also this section about yarn is not needed in the local install with skaffold documentation.

Starting the frontend with yarn is not related to skaffold and it targets mostly frontend developpers in my opinion.
Maybe it could be removed or moved to the substra-frontend README?

Author: Okay thank you for this review @samlesu , for now, I have moved this yarn part to the `Troubleshooting` section!

### Stop Substra

In order to stop Substra, hit `ctrl + c` in each repository. On Ubuntu, if you want to stop the minikube Kubernetes cluster, you can use `minikube stop`.

If you want to restart, you will just need to run again the `skaffold run` command in the 3 repositories.

### Reset Substra

You can reset your installation (if you've used `skaffold run`) with:

```sh
# run from each repositories (hlf-k8s, substra-backend, substra-frontend)
skaffold delete
# or
kubectl rm ns peer-1 peer-2 orderer
# On Ubuntu, to remove all the Kubernetes cluster
minikube delete
```

## Login, password and urls

### Credentials and urls

Once the platform is running, you can sign in to the two organizations using the default development credentials:

org-1:

- API url: `http://substra-backend.node-1.com`
- Frontend url: `http://substra-frontend.node-1.com`
- Username: `node-1`
- Password: `p@$swr0d44`

org-2:

- API url: `http://substra-backend.node-2.com`
- Frontend url: `http://substra-frontend.node-2.com`
- Username: `node-2`
- Password: `p@$swr0d45`

You should find the credentials in the charts: `skaffold.yaml` files or in the `substra-backend/charts/substra-backend/values.yaml`.

> ### Browser extension

Reviewer: It's really minor but I don't think this section should be part of the local intall with skaffold (and even the global documentation).

Using the backend webpage in the browser is in my opinion required only for development purposes and not for end users.

This section could leave only in the substra-backend README?

Author: Okay, noted, for now, I am moving this part to the troubleshooting section as well

            {
                "enabled": true,
                "type": "urls",
                "urlRegex": "http://susbtra-backend.node-1.com"
            }
        ],
        "urlReplacements": [],
        "appendMode": false
    }
]
```

See: <https://github.com/SubstraFoundation/substra-backend#testing-with-the-browsable-api>

### Substra CLI config & login (Mac & Ubuntu)

> Note 1: Substra works on Python 3.6.
>
> Note 2: On Ubuntu, if the command `pip3 install substra` fails, you might need to install `keyrings-alt` with `pip3 install keyrings-alt`; you might even have to do `sudo apt-get remove python3-keyrings.alt python-keyrings.alt`.
>
> Note 3: If you are working inside a virtualized environment, you probably will have execute to `pip3` commands with `sudo`.

Install the CLI:

> Need help to setup a Python Virtual Environment? [Check this out](#python-virtual-environment)

```sh
pip3 install substra
```

Login with the CLI

```sh
# Configuration
substra config -u "node-1" -p 'p@$swr0d44' http://substra-backend.node-1.com

# Login
substra login

# Then you can try
substra list node
# or
substra --help
# or
substra list traintuple
# or
substra get traintuple HASH
```

> This is it, if you the `substra login` command is okay, you're good to go!

Reviewer: ```suggestion
This is it, if the `substra login` command exited successfully, you're good to go!
```

There is a typo in the sentence.

                "type": "urls",
                "urlRegex": "http://substra-backend.node-2.com"
            },
            {
                "enabled": true,
                "type": "urls",
                "urlRegex": "http://susbtra-backend.node-1.com"
            }
        ],
        "urlReplacements": [],
        "appendMode": false
    }
]
```

See: <https://github.com/SubstraFoundation/substra-backend#testing-with-the-browsable-api>

### Substra CLI config & login (Mac & Ubuntu)

> Note 1: Substra works on Python 3+
>
> Note 2: On Ubuntu, if the command `pip3 install substra` fails, you might need to install `keyrings-alt` with `pip3 install keyrings-alt`; you might even have to `sudo apt-get remove python3-keyrings.alt python-keyrings.alt`.
>
> Note 3: If you are working inside a virtualized environment, you probably will have execute to `pip3` commands with `sudo`.

In order to keep your installation of Substra separated from your general Python environement, which is a general Python good practice, it is recommanded to prepare a Python [virtual environment](https://virtualenv.pypa.io/en/latest/). In a new terminal window, please use one of the following method:

> 

Author: In fact, this has been requested and useful for some of our users who were not feeling comfortable with Python venv. Maybe I can move this in the "Futher Resources" section and just had a link like "need help to setup a venv => check this", would it be better?

> 

Reviewer 3: 👍

  - [Helm 2 documentation](https://v2.helm.sh/docs/)
  - [Helm general file structure](https://v2.helm.sh/docs/developing_charts/#the-chart-file-structure)
- Hyperledger Fabric
  - Installation: <https://medium.com/hackernoon/hyperledger-fabric-installation-guide-74065855eca9#c566>
  - Building your first network: <https://hyperledger-fabric.readthedocs.io/en/release-1.4/build_network.html>
- [Awesome Kubernetes list](https://github.com/ramitsurana/awesome-kubernetes#starting-point)
- [Minikube](https://minikube.sigs.k8s.io/) is recommended on Ubuntu but you can also use [Microk8s](https://microk8s.io/).
- Use [Firefox Multi-Account Containers](https://addons.mozilla.org/en-US/firefox/addon/multi-account-containers/) extension to have several simultaneous different logins
- TLDR (*Too Long; Didn't Read*): `sudo apt install tldr` on Ubuntu or `brew install tldr` on Mac
  - `tldr kubectl`
  - `tldr minikube`
  - `tldr helm`
  - `tldr skaffold`
- If you are looking for Python Virtual Environment resource, you might be interested in this post from [Real Python](https://realpython.com/python-virtual-environments-a-primer/).

## Acknowledgements

> This amazing piece of software has been developed and open sourced by [Owkin](https://owkin.com/) and its [terrific developers](https://github.com/SubstraFoundation/substra/graphs/contributors). The repositories are now maintained by [Substra Foundation](https://github.com/SubstraFoundation) and its community. Besides, Substra is really excited to welcome new members, feedbacks and contributions, so please, feel free to get in touch with us!

Author: I already changed it here yesterday from the other review!

### Browser extension

In order to use the backend pages (API) on your browser, you will need to install this extension that will send a special header containing a version:

- [Firefox](https://addons.mozilla.org/en-US/firefox/addon/modheader-firefox/)
- [Chrome](https://chrome.google.com/webstore/detail/modheader/idgpnmonknjnojddfkpgkljpfnnfcklj)

You will then need to:

- Add `Request Header`: `text/html;version=0.0, */*; version=0.0`
- Add `Filters` > `URL Pattern`: `http://susbtra-backend.node-1.com`
- Add `Filters` > `URL Pattern`: `http://susbtra-backend.node-2.com`

TODO: [KO] Or import this to the extension:

```json
[{"title":"Profile 1","hideComment":true,"headers":[   {"enabled":true,"name":"Accept","value":"text/html;version=0.0, */*; version=0.0","comment":""}],"respHeaders":[],"filters":[{"enabled":true,"type":"urls","urlRegex":"http://substra-backend.node-2.com"},{"enabled":true,"type":"urls","urlRegex":"http://susbtra-backend.node-1.com"}],"urlReplacements":[],"appendMode":false}]
```

See: <https://github.com/SubstraFoundation/substra-backend#testing-with-the-browsable-api>

> If you are running the `substra/examples`, you will need to edit the `config.json` as follows:

Reviewer 4: The `config.json` will be removed in the examples in this Pull Request: #108

Author: Oh thank you!

Change from:

```json
{
  "profile_name": "owkin",
  "url": "http://substra-backend.owkin.xyz:8000",
  "username": "substra",
  "password": "p@$swr0d44"
}
```

To:

```json
{
  "profile_name": "owkin",
  "url": "http://substra-backend.node-1.com:80",
  "username": "node-1",
  "password": "p@$swr0d44"
}
```

See: <https://github.com/SubstraFoundation/substra/blob/master/docs/local_install_skaffold.md#login-password-and-urls>

> ### Substra CLI config & login

Reviewer 3: Maybe the docker-compose doc could link to this section for the part on the CLI and SDK ?

- Helm: <https://www.linux.com/tutorials/helm-kubernetes-package-manager/>
- Hyperledger Fabric
  - Installation: <https://medium.com/hackernoon/hyperledger-fabric-installation-guide-74065855eca9#c566>
  - Building your first network: <https://hyperledger-fabric.readthedocs.io/en/release-1.4/build_network.html>
- [Awesome Kubernetes list](https://github.com/ramitsurana/awesome-kubernetes#starting-point)
- [Minikube](https://minikube.sigs.k8s.io/) is recommended but you can also use [Microk8s](https://microk8s.io/).
- Use [Firefox Multi-Account Containers](https://addons.mozilla.org/en-US/firefox/addon/multi-account-containers/) extension to have several simultaneous different logins
- TLDR (*Too Long; Didn't Read*): `sudo apt install tldr`

## 7. Need help

Let's talk:
  - [Create an issue on Github](https://github.com/SubstraFoundation/substra/issues/new)
  - Come to chat with us on [Slack](https://substra-workspace.slack.com/archives/CT54J1U2E)
  - Have a look to the [forum](https://forum.substra.org/)
  - Drop us an [email](mailto:help@substra.org)

>   - Or come meet us *afk* in Paris, Nantes or Limoges!

Reviewer 5: ```suggestion
  - Or come meet us *irl* in Paris, Nantes or Limoges!
```

Author: But, but, what happens behind a keyboard *is* happening *in real life*!

Reviewer 5: Technically yes but the meaning of "afk" is mainly to express that you are "unavailable" ;)

Author: Okay!

```
See: <https://github.com/SubstraFoundation/substra-backend#testing-with-the-browsable-api>

If you are running the `substra/examples`, you will need to edit the `config.json` as follows:

Change from:
```json
{
  "profile_name": "owkin",
  "url": "http://substra-backend.owkin.xyz:8000",
  "username": "substra",
  "password": "p@$swr0d44"
}
```
To:
```json
{
  "profile_name": "owkin",
  "url": "http://substra-backend.node-1.com:80",
  "username": "node-1",
  "password": "p@$swr0d44"
}
```

See: <https://github.com/SubstraFoundation/substra/blob/master/docs/local_install_skaffold.md#login-password-and-urls>

### Substra CLI config & login

> Note: On Ubuntu you might need to install `keyrings-alt` with `pip install keyrings-alt`; you might even have to `sudo apt-get remove python3-keyrings.alt python-keyrings.alt`.

```sh
# Configuration
substra config -u "node-1" -p 'p@$swr0d44' http://substra-backend.node-1.com

# Login
substra login

# Then you can try
substra --help
# or
substra list traintuple
# or

> substra get traintule HASH

Reviewer 4: Looks like you missed a "p" at "traintuple" 🙂

Author: Woupsie, merci, bien vu !

```
See: <https://github.com/SubstraFoundation/substra-backend#testing-with-the-browsable-api>

If you are running the `substra/examples`, you will need to edit the `config.json` as follows:

Change from:
```json
{
  "profile_name": "owkin",
  "url": "http://substra-backend.owkin.xyz:8000",
  "username": "substra",
  "password": "p@$swr0d44"
}
```
To:
```json
{
  "profile_name": "owkin",
  "url": "http://substra-backend.node-1.com:80",
  "username": "node-1",
  "password": "p@$swr0d44"
}
```

See: <https://github.com/SubstraFoundation/substra/blob/master/docs/local_install_skaffold.md#login-password-and-urls>

### Substra CLI config & login

> Note: On Ubuntu you might need to install `keyrings-alt` with `pip install keyrings-alt`; you might even have to `sudo apt-get remove python3-keyrings.alt python-keyrings.alt`.

```sh
# Configuration
substra config -u "node-1" -p 'p@$swr0d44' http://substra-backend.node-1.com

# Login
substra login

# Then you can try
substra --help
# or
substra list traintuple
# or
substra get traintuple HASH
```

## 5. Titanic example

> Note: You don't need to create an account on Kaggle, all the required materials are already provided!

> If you want to ensure that everything is working approprietly, you use the examples provided in the [Substra repository](https://github.com/SubstraFoundation/substra/blob/master/examples/titanic/README.md#testing-our-assets)

Reviewer 2: I don't think we should reference examples in this doc, it distracts from the main focus of just having the platform running.

Reviewer 2: typo: appropriately

Author: Okay!
